### ISSUE TITLE
Is there a way to define that a few properties together must be unique?

### ISSUE THREAD

**Issue by zoechi** [stablekernel/aqueduct#158]
I have a table

`[id, parentTableName, parentId, childTableName, childId]`

and I need a unique index over `parentTableName, parentId, childTableName, childId`

**Comment by zoechi** [comment]
Any workaround for how to define a relationship where the "child" can refer "parents" in different tables is welcome. 
A way to embed an object would also work.

**Comment by joeconwaystk** [comment]
Composite keys aren't supported. They aren't too high of a priority right now, to be honest. Can you clarify what you mean by a child can refer to parents in different tables? Do you mean have multiple foreign keys, but only have one property? The underlying persistent top must have properties the foreign key, but you can create a @managedTransientAttribute property in the ManagedObject subclass that does something like:

```dart
class Child extends ManagedObject<_Child> implements _Child {
  @managedTransientAttribte
  Parent get parent {
    if (parentRelationshipA != null) {
      return parentRelationshipA;
    } else if (parentRelationshipB != null) {
     return parentRelationshipB;
   } ...

}
```

**Comment by zoechi** [comment]
What I mean is that one record in Child belongs to a record in ParentA and another child to a record in ParentB. I'll have a closer look at your suggestion when I'm bach home in a few days.
Thanks a lot.

**Comment by joeconwaystk** [comment]
Got it. So, the way I suggested above requires a column for every possible relationship. See also the thread here: https://github.com/stablekernel/aqueduct/issues/149

**Comment by zoechi** [comment]
I guess I'll go with not modeling a relation and just execute an extra query for now. Not very efficient but there are too many relations to such Child tables from a lot of other tables which makes a column per parent too cumnersome.

> Composite keys aren't supported.

I didn't mean composite keys. Composite keys add a lot of complexity to an ORM and are easy to avoid (if there is no legacy database to be used).

What I ask for is a unique constraint over multiple columns. This is IMHO a mandatory feature.

**Comment by zoechi** [comment]
Back from vacation.
> unique constraint over multiple columns

I think this would only require changes to the code that updates the DB schema and some annotations to define the properties that define a unique constraint or index.

Is there a way (hook) to modify the result of the generated schema update scripts?

**Comment by joeconwaystk** [comment]
@zoechi you mean the Migration script generated by the `aqueduct db` commands? You can add raw SQL commands at anytime during upgrade or seed methods in a Migration subclass. The base class has a PersistentStore, so store.execute(...).

**Comment by zoechi** [comment]
@joeconwaystk sounds promising. I'll have a look.

**Comment by felipefv** [comment]
I find that this functionality (of setting multiple fields on a unique constraint) could help a lot. Thanks!

**Comment by joeconwaystk** [comment]
👍 Will put together a test plan

**Comment by joeconwaystk** [comment]
Adding test cases here

# Data Modeling
- Add ManagedTableAttributes to persistent type with unique list makes instances unique for those columns
- Add ManagedTableAttributes to persistent type with only single element in unique list throws exception, warns to use ManagedColumnAttributes
- Add ManagedTableAttributes to persistent type with empty unique list throws exception
- Add ManagedTableAttributes to persistent type with invalid property in unique list throws exception
  - invalid is "has-"" relationship (belongs to is OK)
  - or property that doesn't exist

# Querying
- Query that inserts row that violates multiple column unique constraint throws 409
- Query that inserts row that satisfies multiple column unique constraint is OK 

# Schema Diff
- No unique constraint -> some unique constraint
- Unique constraint -> minus a key
- Unique constraint -> add a key
- Unique constraint -> swap a key
- Unique constraint -> no unique constraint

# Migration
- Migration fails if constraint can't be met by current values
- Migration succeeds if valid addition
- Migration succeeds if valid change 
- Migration succeeds if valid removal

**Comment by joeconwaystk** [comment]
This is now available on pub, v2.4.0.

**Comment by felipefv** [comment]
Hey, @joeconwaystk ! Do you have any examples on how to use this attribute? I'm getting some erros by using something like "@ManagedTableAttributes(unique: const [#name, #email]);" (which is in the comments on attributes.dart). Thanks in advance!

**Comment by felipefv** [comment]
Just to give a feedback here, I've got an answer on Aqueduct's Slack channel (@joeconwaystk ):

```
@ManagedTableAttributes(uniquePropertySet: const [#first, #second, #etc])
```

Thanks!

**Comment by llakhch** [comment]
is @ManagedTableAttributes available in dart 1.2 ?
